File: 135. Candy/candy.js
/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function (ratings) {
    let len = ratings.length
    let arr = new Array(len).fill(1)

    //left to right
    for (let i = 1; i < len; i++) {              //O(n)
        if (ratings[i] > ratings[i - 1]) {
            arr[i] = arr[i - 1] + 1
        }
    }

    //right tp left
    for (let i = len - 2; i >= 0; i--) {               //O(n)
        if (ratings[i] > ratings[i + 1] && arr[i + 1] >= arr[i]) {
            arr[i] = arr[i + 1] + 1
        }
    }

    return arr.reduce((acc, curr) => acc + curr, 0)     //O(n)
};

/*
- the minimum number of candies
- count the candies must be spread based on rating one from start -> end and another from end -> start
- Total Time Complexity is => O(3 n) = O(n)
*/

File: 135. Candy/candy.cpp
#include <iostream>
#include <vector> 

using namespace std;

class Solution {
    public:
        int candy(vector<int>& ratings) {
            int len = ratings.size();
            int ans = 0;
            vector<int> arr(len,1);
    
            //from left to right
            for(int i=1 ; i<len; i++){
                if(ratings[i] > ratings[i-1]){
                    arr[i] = arr[i-1] + 1;
                }
            }
    
            //from right to left
            for(int i=len-2 ; i>=0; i--){
                if(ratings[i] > ratings[i+1] && arr[i+1] >= arr[i]){
                    arr[i] = arr[i+1] + 1;
                }
            }
    
            for(auto num : arr){
                ans+= num;
            }
    
            return ans;
        }
    };